### PR TITLE
Invoice chanman utility

### PR DESCRIPTION
**Pull request by valentinewallace**
~TODO: tests~

Based on #893

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:2464]: Because its a secret key, we should probably return a reference to it instead of copying the actual secret bytes in memory. Likely to be optimized into the same thing, but its still "best practice".

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:937]: nit: because its a minimum, providing `Some(0)` is the same as `None`. Should we just make it non-`Option`al?

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:943]: Needs features, right, or is just waiting on #876 to land?

**Review by TheBlueMatt — COMMENTED** [review]
Looks good.

File: lightning-invoice/src/lib.rs
Comment (valentinewallace) [reply lightning-invoice/src/lib.rs:943]: Yeah, locally it's based on features. Now getting the sample based on this PR so I can add a test with an invoice I actually know is good...

File: lightning-invoice/src/lib.rs
Comment (valentinewallace) [reply lightning-invoice/src/lib.rs]: I'm fine with `Option`al, switched the param

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:213]: Perhaps we can move `PaymentSecret` into a more low-level module like `ln`. We can reexport it in `channelmanager` to avoid making any cascading changes, if needed.

File: lightning-invoice/src/lib.rs
Comment (jkczyz) [inline lightning-invoice/src/lib.rs:978]: Rather than exposing the secret, maybe `ChannelManager` could provide an interface for signing?

File: lightning-invoice/src/lib.rs
Comment (jkczyz) [inline lightning-invoice/src/lib.rs:913]: Could you move this into a separate utility module akin to `lightning_block_sync::init`? I'd prefer to separate higher-level lightning data structures from lower-level invoice code when possible.

File: lightning-invoice/Cargo.toml
Comment (TheBlueMatt) [inline lightning-invoice/Cargo.toml:21]: Not that it matters, but I think the solution here is to make rust-lightning's _test_utils feature depend on "bitcoin/bitcoinconsensus" instead.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:912]: Looks like we're taking the dependency on rust-bitcoin specifically to take a network here and then convert it to a Currency, can we just take a Currency?

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:934]: I dont think we need to check this? They're additive, doesn't matter if one hop is lower than our final CLTV.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:924]: Maybe use the new constant in one of the later updates of the dependent PRs https://github.com/rust-bitcoin/rust-lightning/pull/893/commits/615ef7d6f88464ed8478a8c8cc0e7582ba66d773

File: lightning-invoice/src/lib.rs
Comment (valentinewallace) [reply lightning-invoice/src/lib.rs:913]: Hm, what do you think the module should be called? `utils` perhaps?

**Comment by TheBlueMatt** [comment]
Can now be rebased after the merge of #893.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [reply lightning/src/ln/channelmanager.rs:213]: Feels a bit odd to separate out the PaymentSecret from the PaymentPreimage/PaymentHash. Maybe they should all be moved in follow-up?

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:213]: I mean its a few extra lines of diff, might as well move them all in one go.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [reply lightning-invoice/src/lib.rs:978]: If we sign anything blindly, it feels like basically the same thing - we'd have to actually have logic knowing what we're signing to avoid it, which would be great, except then we get into lightning depending on lightning-invoice and circular dependencies :/.

File: lightning-invoice/src/lib.rs
Comment (jkczyz) [reply lightning-invoice/src/lib.rs:913]: Yeah, `utils` would be fine.

File: lightning/Cargo.toml
Comment (TheBlueMatt) [inline lightning/Cargo.toml:36]: Oops, we still need this block for `cargo test -p lightning` because that isn't `_test_util` just `test`

Comment (valentinewallace) [reply lightning/Cargo.toml:36]: Ah, thanks that's what I'm looking for

File: lightning-invoice/src/lib.rs
Comment (ariard) [inline lightning-invoice/src/lib.rs:1321]: IIUC this utility returns as routing hints the complete set of our channels, including the non-announced subset one ? 

Sounds pretty harmful for privacy, maybe the utility should have a boolean `all` or otherwise a Vec<short_channel_id> to precise which ones to disclose ?

File: lightning-invoice/src/lib.rs
Comment (ariard) [reply lightning-invoice/src/lib.rs]: Pretty agree with Jeff here, does isolating the utility in its own module resolve the circular dependency ?

Extracting the node secret key from the node will let any malicious entity signs a bunch of random invoices with known hash/preimages. If the victim node is a merchant it opens the way to further social attacks, like contacting the support to claim that a service/good is owed and has not been delivered ?

**Review by ariard — COMMENTED** [review]
Otherwise sounds good to me :)

File: lightning-invoice/src/lib.rs
Comment (jkczyz) [inline lightning-invoice/src/lib.rs:1293]: I was thinking this could be in its own file as well for code cleanliness.

File: fuzz/src/full_stack.rs
Comment (jkczyz) [inline fuzz/src/full_stack.rs:79]: Formatting changes in `full_stack.rs` look undesirable. See commit 5a478c769dc106ea0a0a0747973ef2d0f67288fe.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [reply lightning-invoice/src/lib.rs]: > does isolating the utility in its own module resolve the circular dependency

No, the only solution we could do would be to pull at least `Features` (and probably a few other tiny things) into a new crate which both rust-lightning and lightning-invoice depend on. I'm not super excited about that, but its possible.

> Extracting the node secret key from the node will let any malicious entity signs a bunch of random invoices with known hash/preimages

Yea, we don't really have a concrete story for the node secret's security model right now. We do require it in ChannelManager, so we have to assume anyone with access to the `ChannelManager` can, in principle, do anything they want with the node secret (at least its on the same host/in the same memory space). Its unclear if there's a lot we can do about that, either, we need it to read onions. We could make reading onions a part of the keys interface, but that does seem like an API that is going to change a *ton*.

In general, I think we really need a spec update to let us sign invoices with a different key than onion decryption (etc). cc @devrandom.

Comment (devrandom) [reply lightning-invoice/src/lib.rs]: It would be good not to go further in the direction of passing around the private key.

I think we can have a function `sign_invoice` on `KeysInterface` with the following arguments:

- hrp
- data

and they could be either `Vec<u5>` or `Vec<u8>`.

If the developer actually wants to validate inside `KeysInterface`, they will have to re-parse the invoice.  I think that's fine, since you likely go across process boundaries when signing in the general case (hardware wallet / HSM / hardened signing server).

As to the node secret in other contexts - I think it boils down to signing things and computing shared secrets.  I think both of these can be behind `KeysInterface` (or a smaller trait that can be gotten from it).  I'm happy to try implementing there's doubt.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [reply lightning-invoice/src/lib.rs:1321]: I think this is the right default to this - in most cases if we're a "client" node we only have one or two channels anyway (and in non-"client" cases we have several channels that are public), and the "real" fix is to generate random short channel ids.

**Comment by valentinewallace** [comment]
Oops, need to switch `sign_invoice` to return a `Result`. Otherwise I think this is roughly what it's gonna look like.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [reply lightning-invoice/src/lib.rs:1293]: Probably also should be pub :).

Comment (TheBlueMatt) [reply lightning-invoice/src/lib.rs:1293]: Probably also should be pub in some way :).

File: lightning-invoice/src/lib.rs
Comment (jkczyz) [inline lightning-invoice/src/lib.rs:1392]: The signing API is designed to allow custom errors. So the error type here should be what is returned by `sign_invoice`. Something like:

```rust
let signed_raw_invoice = raw_invoice.sign(|_| keys_manager.sign_invoice(invoice_preimage))?;
```

 But may have to convert to a `SignOrCreationError`.

File: fuzz/src/full_stack.rs
Comment (TheBlueMatt) [reply fuzz/src/full_stack.rs:80]: Looks like rustfmt accidentally got run across this file?

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:1293]: nit indentation here.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:1309]: This shouldnt be dead code if its publicly-exposed.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:1308]: Probably useful to describe when you'd want to use this and why.

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:1312]: Should be no reason to need a dyn here - can we just add a template parameter?

File: lightning-invoice/src/lib.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/lib.rs:1310]: May be useful to call this `Invoice_from_...` or so. In context I guess it makes sense but if its just in a module on its own it would be nice to reference the thing being created. Do you agree @jkczyz?

Comment (jkczyz) [reply lightning-invoice/src/lib.rs:1310]: Maybe `create_invoice_from_channelmanager` since using a `from_` prefix tends to imply a conversion.

File: lightning-invoice/src/lib.rs
Comment (valentinewallace) [reply lightning-invoice/src/lib.rs:1313]: Hm so `Invoice::from_signed` returns ` SemanticError`, which is a bit tricky to translate to a `CreationError`

Comment (jkczyz) [reply lightning-invoice/src/lib.rs:1313]: Exactly. This would indicate a bug on our end so we should `unwrap` or `expect` there.

**Comment by valentinewallace** [comment]
Gonna have lunch then come back to this

File: lightning/src/chain/keysinterface.rs
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:400]: `DecodeError` definitely feels pretty strange here - I suppose in principle users may be decoding so that's a part of the possible errors, but users are also signing, and you could imagine something like "hard ware wallet unplugged" as an error. I guess we could do `()` for the error type  here like we do a few others - ultimately we're just gonna pass the user their own error back again.

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:2466]: This should be dropped now.

File: lightning-invoice/src/utils.rs
Comment (jkczyz) [inline lightning-invoice/src/utils.rs:100]: bad indentation

File: lightning/src/chain/keysinterface.rs
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:1064]: You can simplify this a good bit (plus then drop the `Message` import above.
```suggestion
		Ok(self.secp_ctx.sign_recoverable(&hash_to_message!(&Sha256::hash(&invoice_preimage)), &self.get_node_secret()))
```

File: lightning-invoice/src/utils.rs
Comment (jkczyz) [inline lightning-invoice/src/utils.rs:101]: Type declaration is probably unnecessary now.

File: lightning-invoice/src/utils.rs
Comment (jkczyz) [inline lightning-invoice/src/utils.rs:105]: nit:

```rust
match signed_raw_invoice {
	Ok(inv) => Ok(Invoice::from_signed(inv).unwrap()),
	Err(e) => Err(SignOrCreationError::SignError(e)),
}
```

File: lightning-invoice/src/utils.rs
Comment (jkczyz) [inline lightning-invoice/src/utils.rs:23]: Rename `currency` to `network` and consider moving it before `amt_msat`.

**Review by TheBlueMatt — COMMENTED** [review]
Aside from Jeff's comments and https://github.com/rust-bitcoin/rust-lightning/pull/895#discussion_r623219246 this LGTM.

File: lightning-invoice/src/lib.rs
Comment (ariard) [reply lightning-invoice/src/lib.rs]: Agree for the "client" node but for a big hub you will disclose all your private channels with spoke *if* you're also a merchant which is quite the case of a bunch of such nodes. Though opening an issue for now, we'll adjust in function of users feedback.

Comment (TheBlueMatt) [reply lightning-invoice/src/lib.rs]: Right, we ultimately need some heuristic of "we have > 5 public channels, we shouldn't include any route hints at all", probably.

File: lightning-invoice/src/lib.rs
Comment (jkczyz) [inline lightning-invoice/src/lib.rs:750]: nit: leave as tab

**Review by TheBlueMatt — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]

File: lightning-invoice/src/utils.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:1]: Lol, oops, needs a `pub mod utils` in lib.rs or this doesn't even compile.

**Review by TheBlueMatt — CHANGES_REQUESTED** [review]

**Comment by valentinewallace** [comment]
Gah my bad, fixed

File: lightning-invoice/src/utils.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:1]: In order for this to be a doc comment it needs to be "//!" not "///"

File: lightning-invoice/src/utils.rs
Comment (TheBlueMatt) [inline lightning-invoice/src/utils.rs:24]: This doesn't compile now.

File: lightning-invoice/src/utils.rs
Comment (valentinewallace) [reply lightning-invoice/src/utils.rs]: TIL

**Pull request merged**
